base_layer/core/src/proof_of_work/cuckaroo_pow.rs

```diff
@@ -0,0 +1,476 @@
+    proof_of_work::{siphash::siphash_block, Difficulty},
+};
+
+#[derive(Debug, Error, PartialEq, Eq)]
+pub enum CuckarooVerificationError {
+    #[error("PoW data contains non-zero padding")]
+    PowDataContainsNonZeroPadding,
+    #[error("PoW data is too short")]
+    PowDataTooShort,
+    #[error("Block header has an invalid PoW algorithm for Cuckaroo")]
+    BlockHeaderInvalidPowAlgorithm,
+    #[error("Nonce is too large")]
+    NonceTooLarge,
+    #[error("Nonces not ascending")]
+    NoncesNotAscending,
+    #[error("Endpoints don't match up")]
+    EndpointsDontMatch,
+    #[error("CycleTooShort")]
+    CycleTooShort,
+    #[error("CycleTooLong")]
+    CycleTooLong,
+    #[error("Edge already visited")]
+    EdgeAlreadyVisited,
+    #[error("Node has more than two edges")]
+    NodeHasMoreThanTwoEdges,
+    #[error("Cycle does not end at start")]
+    CycleDoesNotEndAtStart,
+    #[error("Cycle did not use all edges")]
+    CycleDidNotUseAllEdges,
+}
+
+pub fn cuckaroo_result(
+    header: &BlockHeader,
+    required_cycle_length: usize,
+    edge_bits: u8,
+) -> Result<Vec<u8>, anyhow::Error> {
+    let pow = header.pow.to_bytes();
+
+    // First byte must be 3 for Cuckaroo
+    if pow[0] != 3 {
+        return Err(anyhow::anyhow!(
+            CuckarooVerificationError::BlockHeaderInvalidPowAlgorithm
+        ));
+    }
+    let mut hasher = Blake2bVar::new(32).expect("Could not create Blake2bVar hasher");
+    hasher.update(&header.nonce.to_le_bytes());
+    hasher.update(header.mining_hash().as_slice());
+    let mut blob = Vec::with_capacity(32);
+    hasher.finalize_variable(&mut blob)?;
+
+    let pow_data = &pow[1..];
+    // Data after <required_cycle_length * edge_bits> is padding, it must be zero
+    for &byte in &pow_data[required_cycle_length * edge_bits as usize / 8..] {
+        if byte != 0 {
+            return Err(anyhow::anyhow!(
+                CuckarooVerificationError::PowDataContainsNonZeroPadding
+            ));
+        }
+    }
+
+    let nonces = unpack_nonces(pow_data, edge_bits, required_cycle_length)?;
+    // There might be extra padding at  the end of the nonces.
+    for n in &nonces[required_cycle_length..] {
+        if *n != 0 {
+            return Err(anyhow::anyhow!(CuckarooVerificationError::PowDataTooShort));
+        }
+    }
+    let siphash_keys = [
+        u64::from_le_bytes(blob[0..8].try_into().unwrap()),
+        u64::from_le_bytes(blob[8..16].try_into().unwrap()),
+        u64::from_le_bytes(blob[16..24].try_into().unwrap()),
+        u64::from_le_bytes(blob[24..32].try_into().unwrap()),
+    ];
+    // // Generate the hasher.
+    verify(&siphash_keys, &nonces, required_cycle_length, edge_bits)?;
+
+    // nonces must be sorted, so we just hash them
+    let mut hasher = Blake2bVar::new(32).expect("Could not create Blake2bVar hasher");
+
+    for nonce in &nonces {
+        hasher.update(&nonce.to_le_bytes());
+    }
+    let mut res = Vec::with_capacity(32);
```

Comment:
ditto `fn finalize_variable`
```suggestion
    let mut res = vec![0u8; 32];
```